Add Packet provider and first profile

deitch opened the pull request:
This PR adds a functional packet provider and an ubuntu based first profile. 

The `AUTHORS.md` was updated automatically when `make all` ran, not sure about all of the changes.

/cc @vielmetti

vielmetti commented:
Looks like Travis failed in "make check-headers"

deitch commented:
Oh yes, I see it. Hold on.

deitch commented:
Updated.

xmudrii commented on cloud/packet/packetSDK/sdk.go:
```diff
@@ -0,0 +1,43 @@
+// Copyright © 2017 The Kubicorn Authors
```
We would appreciate tests here. [Example of Happy/Sad tests can be found here](https://github.com/kris-nova/kubicorn/blob/master/cloud/digitalocean/godoSdk/sdk_test.go).

deitch replied:
Sure thing. Will add.

xmudrii commented on cloud/packet/public/resources/device.go:
```diff
@@ -0,0 +1,368 @@
+		newResource.Tags = r.Tags
+	}
+
+	fmt.Println(immutable)
```
I don't believe that `fmt` is meant to be here.

deitch replied:
Ugh! One issue I had when working through it, threw that in because of line and timing issues and never deleted. Nice catch...

xmudrii commented on cloud/packet/public/resources/device.go:
```diff
@@ -0,0 +1,368 @@
+	logger.Debug("device.Actual newResource %v", newResource)
+
+	newCluster := r.immutableRender(newResource, immutable)
+	fmt.Println(newCluster)
```
Same here.

xmudrii commented on cloud/packet/public/resources/device.go:
```diff
@@ -0,0 +1,368 @@
+	expectedResource.ProjectID = projectID
+	logger.Debug("device.Apply expected vs actual %v %v", *expectedResource, *actualResource)
+	j, _ := json.Marshal(*expectedResource)
+	fmt.Println(string(j))
```
Same here.

xmudrii commented on cloud/packet/public/resources/device.go:
```diff
@@ -0,0 +1,368 @@
+	j, _ := json.Marshal(*expectedResource)
+	fmt.Println(string(j))
+	j, _ = json.Marshal(*actualResource)
+	fmt.Println(string(j))
```
Same here.

xmudrii commented on cloud/packet/public/resources/device.go:
```diff
@@ -0,0 +1,368 @@
+		Type:             deleteResource.Type,
+		Location:         deleteResource.Location,
+		BootstrapScripts: deleteResource.BootstrapScripts,
+		Count:            0,
```
Is `count: 0` indeed to be zero?

deitch replied:
I think so, but am not sure. This is at the end of a successful `Delete()` operation. At this point, is the returned `Resource` expected to return:

* the _actual_ state of the cluster, In which case it should be `0`?
* the _configured_ state of the cluster, in which case it should be what was configured originally, but we happen to have `0` because they were deleted, in which case it should be `1` (or whatever was in the profile)?

Which one should be returned here?

xmudrii replied:
Not really sure, but as far as I know, `Delete` returns actual state. Actual state is `Count: 0`, so it should stay zero. Thanks for checking with me.

xmudrii commented on cloud/packet/public/resources/device.go:
```diff
@@ -0,0 +1,368 @@
+		newResource.ProjectID = project.ID
+
+	} else {
+		newResource.Count = 0
```
Same here, is count expected to be zero?

deitch replied:
Same question as with the delete. If `Actual()` runs and discovers that there are no devices that match the tag (e.g. we looked for `master` nodes and discovered there are none), then `len(valid)` will be `0`. In that case, should we not return how many we found, which is `0`?

xmudrii replied:
This one makes sense, so leave it as it is.

xmudrii commented:
@deitch Thank you so much! We really appreciate it!
I've left you some comments that needs to be addressed before we can merge this. If you have any questions, ping me here on Slack.

Once we get this fix, I'll happily merge this PR.

deitch commented:
Removed all of those (awful) `fmt.Println()`. :-)

deitch commented:
Added the requested test. Also updated the documentation to include a "packet-walkthrough.md" like the ones for Google, DO, AWS, etc.

I think this is it, but if I missed anything, let me know.

xmudrii commented:
Thank you so much for the hard work here!! 🎉 :shipit:
